common/web/lm-worker/src/main/model-compositor.ts

```diff
@@ -380,83 +404,83 @@ export default class ModelCompositor {
     // Assumption:  Duplicated 'displayAs' properties indicate duplicated Suggestions.
     // When true, we can use an 'associative array' to de-duplicate everything.
     let suggestionDistribMap: {[key: string]: CorrectionPredictionTuple} = {};
-    let currentCasing: CasingForm = null;
-    if(lexicalModel.languageUsesCasing) {
-      currentCasing = this.detectCurrentCasing(postContext);
-    }
 
-    let baseWord = this.wordbreak(context);
+    const keyed = (text: string) => this.lexicalModel.toKey ? this.lexicalModel.toKey(text) : text;
+    const keyCased = (text: string) => this.lexicalModel.applyCasing ? this.lexicalModel.applyCasing('lower', text) : text;
+    const keyedPrefix = keyed(truePrefix);
+    const lowercasedPrefix = keyCased(truePrefix);
 
     // Deduplicator + annotator of 'keep' suggestions.
     for(let tuple of rawPredictions) {
-      const prediction = tuple.prediction.sample;
-      const prob = tuple.totalProb;
-
-      // Combine duplicate samples.
-      let displayText = prediction.displayAs;
-      let preserveAsKeep = displayText == keepOptionText;
-
-      // De-duplication should be case-insensitive, but NOT
-      // diacritic-insensitive.
-      if(this.lexicalModel.languageUsesCasing) {
-        preserveAsKeep = preserveAsKeep || displayText == this.lexicalModel.applyCasing('lower', keepOptionText);
+      if(currentCasing && currentCasing != 'lower') {
+        this.applySuggestionCasing(tuple.prediction.sample, basePrefix, currentCasing);
       }
 
-      if(preserveAsKeep) {
-        // Preserve the original, pre-keyed version of the text.
-        if(!keepOption) {
-          let baseTransform = prediction.transform;
+      // Don't set it unnecessarily; this can have side-effects in some automated tests.
+      if(inputTransform.id !== undefined) {
+        tuple.prediction.sample.transformId = inputTransform.id;
+      }
 
-          let keepTransform = {
-            insert: keepOptionText,
-            deleteLeft: baseTransform.deleteLeft,
-            deleteRight: baseTransform.deleteRight,
-            id: baseTransform.id
-          }
+      const predictedWord = this.wordbreak(models.applyTransform(tuple.prediction.sample.transform, context));
 
-          let intermediateKeep = models.transformToSuggestion(keepTransform, prob);
-          keepOption = this.toAnnotatedSuggestion(intermediateKeep, 'keep',  models.QuoteBehavior.noQuotes);
+      // Is the suggestion an exact match (or, "similar enough") to the
+      // actually-typed context?  If so, we wish to note this fact and to
+      // prioritize such a suggestion over suggestions that are not.
+      if(keyed(tuple.correction.sample) == keyedPrefix) {
+        if(predictedWord == truePrefix) {
+          tuple.matchLevel = SuggestionSimilarity.exact;
+          keepOption = this.toAnnotatedSuggestion(tuple.prediction.sample, 'keep',  models.QuoteBehavior.noQuotes);
```

Comment:
It would be good to add a comment for this and the next 3 lines of code because it's not entirely obvious what's happening here, even with the comments on ll.426-428